### PR TITLE
Creating Issue Reports for Flaky Test Failures in Gradle Check

### PR DESCRIPTION
**Pull request by prudhvigodithi**
### Description
Gradle Check flaky test failures issue creation

### Issues Resolved
Part of https://github.com/opensearch-project/OpenSearch/issues/13950

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by prudhvigodithi** [comment]
Adding @dblock since a similar setup was done in past for code inside [src/jenkins](https://github.com/opensearch-project/opensearch-build-libraries/tree/main/src/jenkins). I have added a package `gradlecheck` that has the code which can be used by the library `gradleCheckFlakyTestIssueCreate.groovy` and jenkinsfile can simply load and call this library. The only change is I did not  add `implements Serializable` also unlike each class part of [src/jenkins](https://github.com/opensearch-project/opensearch-build-libraries/tree/main/src/jenkins) the code inside `src/gradlecheck` is used only only `gradleCheckFlakyTestIssueCreate` library, I have done this way to make sure the code is organized following OOP methods rather than just dumping all the lines under one library inside `var/gradleCheckFlakyTestIssueCreate`. 

Having this now the unit tests I have used `mockito-core` for testing the code inside `tests/gradlecheck` and works fine (example `./gradlew test --tests=OpenSearchMetricsQueryTest`), but when invoked with existing tests the old tests times out with the following error.

```
2024-06-11T06:44:05.566+0000 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Lock acquired on daemon addresses registry.
2024-06-11T06:44:05.566+0000 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Releasing lock on daemon addresses registry.
2024-06-11T06:44:05.567+0000 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Waiting to acquire shared lock on daemon addresses registry.
2024-06-11T06:44:05.567+0000 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Lock acquired on daemon addresses registry.
2024-06-11T06:44:05.567+0000 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Releasing lock on daemon addresses registry.
```

Just deleting the directory `src/gradlecheck` and running the existing test example `./gradlew clean test --tests=TestUploadTestResults -info -Ppipeline.stack.write=true` again moves forward without any timeout error. 

I'm wondering why just adding one extra package `src/gradlecheck` and testing this with existing test suite is not possible ? @gaiksaya @peterzhuamazon @rishabh6788 @getsaurabh02

**Comment by prudhvigodithi** [comment]
Coming from my above [comment](https://github.com/opensearch-project/opensearch-build-libraries/pull/436#issuecomment-2161461640) the hanging tests issue is due to inheritance used in the `src/gradlecheck` code files, even though this works in runtime and able to complete the end requirement, this hangs during testing. However with the same goal of using `src/gradlecheck` I have modified the code so that its compatible during testing and does not break the existing tests still honoring  fetching the required classes from `src/gradlecheck` for the `gradleCheckFlakyTestChecker` library. 

I have also modified the library `vars/createGithubIssue` that will be used in `gradleCheckFlakyTestChecker` to create/update the issue body with `--body-file` and also allow to edit the issue body (not just comment), this change broke majority of the tests that use `createGithubIssue` directly or indirectly.

**Comment by prudhvigodithi** [comment]
The CI checks are now green and library is tested as jenkins run, sample issues that are created https://github.com/prudhvigodithi/OpenSearch/issues/25 https://github.com/prudhvigodithi/OpenSearch/issues/24.

**Comment by rishabh6788** [comment]
> The CI checks are now green and library is tested as jenkins run, sample issues that are created [prudhvigodithi/OpenSearch#25](https://github.com/prudhvigodithi/OpenSearch/issues/25) [prudhvigodithi/OpenSearch#24](https://github.com/prudhvigodithi/OpenSearch/issues/24).

I see that there are two comments, if I remember correctly wasn't it supposed to be just one comment that would keep getting updated or description?
Also, I feel the way data is being presented can be improved.

**Comment by prudhvigodithi** [comment]
> I see that there are two comments, if I remember correctly wasn't it supposed to be just one comment that would keep getting updated or description?
> Also, I feel the way data is being presented can be improved.

Ya I have removed the `issueEdit` option to test the default behaviour (comment on an open issue rather than editing the body) of the `createGithubIssue` library. I have added it back.

The markdown table is coming from the approval https://github.com/opensearch-project/OpenSearch/issues/13950#issuecomment-2155221675, this has all the information per commit, associated PR, jenkins build and failed tests.
We can always improve based on the feedback once we see this in action by start creating the issues.

File: vars/createGithubIssue.groovy
Comment (rishabh6788) [inline vars/createGithubIssue.groovy:27]: In case there are multiple issues returned then is this picking up the top result?

File: src/gradlecheck/FetchPostMergeFailedTestName.groovy
Comment (rishabh6788) [inline src/gradlecheck/FetchPostMergeFailedTestName.groovy:128]: nit: missing space after comma.

File: src/gradlecheck/OpenSearchMetricsQuery.groovy
Comment (rishabh6788) [inline src/gradlecheck/OpenSearchMetricsQuery.groovy:30]: just for my knowledge, what is 'script.sh'?

**Comment by rishabh6788** [comment]
One thought I have is once the workflow has created an issue, and someone works on it, may be updates the comment or description and strikes-through a few failing tests that have been fixed. 
Now when the workflow runs once again, wouldn't it override the edits made by operator? 
Are we assuming that operator would only be adding new comments on the issue and not editing the comment/description added by with workflow?

File: vars/createGithubIssue.groovy
Comment (prudhvigodithi) [reply vars/createGithubIssue.groovy]: Yes Risab that was the idea for this existing library `createGithubIssue`, the logic was with same title for all the AUTOCUT issues, there will be only one issue created for one scenario.

File: src/gradlecheck/OpenSearchMetricsQuery.groovy
Comment (prudhvigodithi) [reply src/gradlecheck/OpenSearchMetricsQuery.groovy:30]: The `script` is a jenkins context and will be injected to the library during runtime. Without the jenkins context the native groovy does have the `sh` inbuit.

**Comment by prudhvigodithi** [comment]
> One thought I have is once the workflow has created an issue, and someone works on it, may be updates the comment or description and strikes-through a few failing tests that have been fixed.
> Now when the workflow runs once again, wouldn't it override the edits made by operator?
> Are we assuming that operator would only be adding new comments on the issue and not editing the comment/description added by with workflow?

AFAIK only maintainers/admins's can edit the issue created by external users and I dont think one can edit the issue created by other user.

File: src/gradlecheck/FetchPostMergeFailedTestName.groovy
Comment (prudhvigodithi) [reply src/gradlecheck/FetchPostMergeFailedTestName.groovy]: Added thanks

File: src/gradlecheck/FetchPostMergeFailedTestClass.groovy
Comment (rishabh6788) [inline src/gradlecheck/FetchPostMergeFailedTestClass.groovy:22]: except markdown class, every class has the same constructor and `get_query` methods. Would it be better to have a parent/super class or interface, if supported in groovy class, and have these method defined there and other classes just inherit it and implement in its own way. Method like `getPostMergeFailedTestClass` can have a generic name and implemented in all the sub-classes. 
Not a blocker for me.

Comment (prudhvigodithi) [reply src/gradlecheck/FetchPostMergeFailedTestClass.groovy:22]: That was my initial implementation rishab, take a look at this [commit](https://github.com/opensearch-project/opensearch-build-libraries/commit/b8462b4d890cc2a218bab88924f006d5a1ff0cbd#diff-2087681af514d648a0a6740a4b91375803a761294436f8d010934d4ea7ffdd8d) which has the `extends` and `super` implementation, the jenkins pipeline works but this  `extends` and `super` implementation ran into issues with running the tests (both old and new tests part of the PR) part of this repo, please check my [comment](https://github.com/opensearch-project/opensearch-build-libraries/pull/436#issuecomment-2162026372) here,

File: src/gradlecheck/OpenSearchMetricsQuery.groovy
Comment (prudhvigodithi) [reply src/gradlecheck/OpenSearchMetricsQuery.groovy:30]: Adding more to this, to run a script in groovy the ideal way to use is `.execute()`, but running `.execute()` in jenkins is blocked by the script security plugin, hence the ideal was is to use `sh` given by jenkins and for `sh` to work in libraries it needs jenkins context.

**Comment by prudhvigodithi** [comment]
FYI I have added a comment to an open issue https://github.com/jenkinsci/JenkinsPipelineUnit/issues/595#issuecomment-2164014788 on why we cant test the library directly `gradleCheckFlakyTestChecker` that has the external classes. Here for our case the `gradleCheckFlakyTestChecker` just uses the code part of `src/gradlecheck`, for all the code part of `src/gradlecheck` is tested and covered in `tests/gradlecheck` (Refer the CodeCov https://github.com/opensearch-project/opensearch-build-libraries/pull/436#issuecomment-2161461640).

**Review by rishabh6788 — APPROVED** [review]
Approving assuming all search queries are working as expected.

**Comment by prudhvigodithi** [comment]
Thanks @rishabh6788 just pushed a commit change the repo from my fork to `repoUrl: "https://github.com/opensearch-project/OpenSearch",`.

**Pull request merged**
